webserver.py

```diff
@@ -41,61 +42,178 @@ def get_conn():
-            if game_state.get('plot' + str(i) + '_' + str(j)) is None:
-                game_state['plot' + str(i) + '_' + str(j)] = {'seedType': 0, 'sowTime': 0, 'locked': 1}
-            if get_plot(game_state, i, j)['seedType'] != 0:
-                total_crops_planted += 1
-    for i in range(2):
-        for j in range(2):
-            get_plot(game_state, i, j)['locked'] = 0
-    if total_seed_count == 0 and total_crops_planted == 0:
-        if game_state['resources']['cash'] < GAME_CONFIG['starting_resources']['cash']:
-            game_state['resources']['cash'] = GAME_CONFIG['starting_resources']['cash']
-    if game_state.get('recipes') is None:
-        game_state['recipes'] = []
-    return game_state
-
-
-def save_state(slug, game_state):
-    my_conn = get_conn()
-    json_game_state = json.dumps(game_state, indent=4)
-    with my_conn:
-        with my_conn.cursor() as cursor:
-            cursor.execute('INSERT INTO game_states (slug, game_state) VALUES (%s, %s) '
-                           'ON CONFLICT (slug) DO UPDATE SET game_state=%s;',
-                           [slug, json_game_state, json_game_state])
+class GameState(object):
+    def __init__(self, json_data):
+        self.data = json_data
+
+    @staticmethod
+    def new(slug, password):
+        game_state = GameState.load(slug, password)
+        if game_state:
+            raise Forbidden("Username already taken")
+        else:
+            game_state = GameState({
+                'resources': {},
+                'slug': slug,
+                'password': password,
+                'unlockCount': 0,
+                'seedCounts': {},
+                'recipes': ['a']
+            })
+        return game_state.initialize()
+
+    @classmethod
+    def load(cls, slug, password):
+        my_conn = get_conn()
+        with my_conn:
+            with my_conn.cursor() as cursor:
+                cursor.execute('SELECT game_state FROM game_states WHERE slug = %s;', [slug])
+                row = cursor.fetchone()
+                if not row:
+                    return None
+                game_state = GameState(row[0])
+                game_state.check_password(password)
+                return game_state.initialize()
+
+    def save(self, slug):
+        my_conn = get_conn()
+        json_game_state = json.dumps(self.data, indent=4)
+        with my_conn:
+            with my_conn.cursor() as cursor:
+                cursor.execute('INSERT INTO game_states (slug, game_state) VALUES (%s, %s) '
+                               'ON CONFLICT (slug) DO UPDATE SET game_state=%s;',
+                               [slug, json_game_state, json_game_state])
+
+    # add missing seed counts, resources, and plots to game_state
+    # unlock first four plots
+    # check to never run out of money
+    def initialize(self):
+        for seed_id in GAME_CONFIG['seeds']:
+            if self.data['seedCounts'].get(seed_id) is None:
```

Comment:
You're right-- this is getting a little ahead of ourselves (you'll do this in #77) but this is a good place to talk about it.  Don't worry about changing it in this PR, but:

this chunk of code:
```
        for seed_id in GAME_CONFIG['seeds']:
            if self.data['seedCounts'].get(seed_id) is None:
                self.data['seedCounts'][seed_id] = 0
```
only exists to initialize the game_state to 0s when a new seed is introduced.  But we can make a rule that says, "when a seed count doesn't exist, we consider it a 0":

```
def get_seed_count(self, seed_id):
    count = self.data['seedCounts'].get(seed_id)
    if count is None:
        return 0
    return count
```

now all your `self.data['seedCounts'].get(seed_id)` calls can just be `self.get_seed_count(seed_id)`.  More importantly, you don't run the risk that if you ever write code where `initialize` fails to get called at the right time, this part won't break-- you'll never end up with a `None` seed_count